pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/XPathPanelController.java

```diff
@@ -112,6 +124,85 @@ public void initialize(URL location, ResourceBundle resources) {
+                               String input = xpathExpressionArea.getText().substring(
+                                   indexOfSlash + 1, m.getInsertionEnd());
+                               m.getRemovalEnd();
+                               return Tuples.t(indexOfSlash, input);
+                           })
+                           .filter(t -> t._2.matches("[a-zA-Z]+"))
+                           .subscribe(e -> autoComplete(e._1, e._2));
+
+        xpathExpressionArea.addEventHandler(KeyEvent.KEY_PRESSED, t -> {
+            if ((t.isControlDown() && t.getCode().isWhitespaceKey()) || t.getCode() == KeyCode.BACK_SPACE) {
+                autoCompletePopup.show(xpathExpressionArea, xpathExpressionArea.getCharacterBoundsOnScreen(positionFrom, positionTo)
+                                                                               .get()
+                                                                               .getMaxX(), xpathExpressionArea
+                                           .getCharacterBoundsOnScreen(positionFrom, positionTo).get().getMaxY());
+            }
+        });
+
+        xpathExpressionArea.addEventHandler(KeyEvent.KEY_PRESSED, t -> {
+            if (t.getCode() == KeyCode.ESCAPE) {
+                autoCompletePopup.hide();
+            }
+        });
+
+    }
+
+    private void autoComplete(int slashPosition, String input) {
+
+        XPathSuggestions xPathSuggestions = new XPathSuggestions(parent.getLanguageVersion().getLanguage());
+        List<String> suggestions = xPathSuggestions.getXPathSuggestions(input.trim());
+
+        List<CustomMenuItem> resultToDisplay = new ArrayList<>();
+        if (suggestions.size() > 0) {
+
+            for (int i = 0; i < suggestions.size() && i < 5; i++) {
+                final String searchResult = suggestions.get(i);
+
+                Label entryLabel = new Label();
+                entryLabel.setGraphic(Style.highlight(suggestions.get(i), input));
+                entryLabel.setPrefHeight(5);
+                CustomMenuItem item = new CustomMenuItem(entryLabel, true);
+                resultToDisplay.add(item);
+
+                item.setOnAction(e -> {
+                    xpathExpressionArea.replaceText(slashPosition + 1, slashPosition + input.length(), searchResult);
+                    autoCompletePopup.hide();
+                });
+            }
+        }
+        autoCompletePopup.getItems().setAll(resultToDisplay);
+        autoCompletePopup.show(xpathExpressionArea, xpathExpressionArea.getCharacterBoundsOnScreen(positionFrom, positionTo)
+                                                                       .get()
+                                                                       .getMaxX(), xpathExpressionArea
+                                   .getCharacterBoundsOnScreen(positionFrom, positionTo).get().getMaxY());
+    }
+
+    private int reverseStringSearch(char search, String input, int begin) {
```

Comment:
isn't this whole method equivalent to `input.lastIndexOf(search, begin - 1)`?

Reply:
Oh yes indeed.